kasa/xortransport.py

```diff
@@ -0,0 +1,225 @@
+
+    async def _execute_send(self, request: str) -> Dict:
+        """Execute a query on the device and wait for the response."""
+        assert self.writer is not None  # noqa: S101
+        assert self.reader is not None  # noqa: S101
+        debug_log = _LOGGER.isEnabledFor(logging.DEBUG)
+        if debug_log:
+            _LOGGER.debug("%s >> %s", self._host, request)
+        self.writer.write(XorEncryption.encrypt(request))
+        await self.writer.drain()
+
+        packed_block_size = await self.reader.readexactly(self.BLOCK_SIZE)
+        length = _UNSIGNED_INT_NETWORK_ORDER.unpack(packed_block_size)[0]
+
+        buffer = await self.reader.readexactly(length)
+        response = XorEncryption.decrypt(buffer)
+        json_payload = json_loads(response)
+        if debug_log:
+            _LOGGER.debug("%s << %s", self._host, pf(json_payload))
+
+        return json_payload
+
+    async def close(self) -> None:
+        """Close the connection."""
+        writer = self.writer
+        self.close_without_wait()
+        if writer:
+            with contextlib.suppress(Exception):
+                await writer.wait_closed()
+
+    def close_without_wait(self) -> None:
+        """Close the connection without waiting for the connection to close."""
+        writer = self.writer
+        self.reader = self.writer = None
+        if writer:
+            writer.close()
+
+    async def reset(self) -> None:
+        """Reset the transport."""
+        await self.close()
+
+    async def send(self, request: str) -> Dict:
+        """Send a message to the device and return a response."""
+        #
+        # Most of the time we will already be connected if the device is online
+        # and the connect call will do nothing and return right away
+        #
+        # However, if we get an unrecoverable error (_NO_RETRY_ERRORS and
+        # ConnectionRefusedError) we do not want to keep trying since many
+        # connection open/close operations in the same time frame can block
+        # the event loop.
+        # This is especially import when there are multiple tplink devices being polled.
+        try:
+            await self._connect(self._timeout)
+        except ConnectionRefusedError as ex:
+            await self.reset()
+            raise SmartDeviceException(
+                f"Unable to connect to the device: {self._host}:{self._port}: {ex}"
+            ) from ex
+        except OSError as ex:
+            await self.reset()
+            if ex.errno in _NO_RETRY_ERRORS:
+                raise SmartDeviceException(
+                    f"Unable to connect to the device:"
+                    f" {self._host}:{self._port}: {ex}"
+                ) from ex
+            else:
+                raise RetryableException(
+                    f"Unable to connect to the device:"
+                    f" {self._host}:{self._port}: {ex}"
+                ) from ex
+        except Exception as ex:
+            await self.reset()
```

Comment:
```suggestion
            await self.close()
```

Reply:
But I have to have it in the `BaseTransport` class because the `protocol` is not aware of these details.

Reply:
How about 
https://github.com/python-kasa/python-kasa/pull/646/files#r1465502942